### PR TITLE
ASR BPE Patches

### PR DESCRIPTION
**Pull request by titu1994**
# Changelog
- Patches signature change in ASR data preprocessor
- Patches WPE signature to allow transcription
- Registers the tokenizer objects (vocab and potentially the .model file) for restoration from .nemo object
- Patches an issue with how number of tokens were calculated (and passed) to the CTC model. Now special tokens will be added to the vocabulary of the model
- Add concrete transcript function based on base ASR model
- Refactors restore_from() so that it changes directory to the temp directory. This is necessary for artifact restoration from .nemo files.
- Add basic instantiation and serialization/deserialization from .nemo tests for BPE models

### CODE REVIEW

File: nemo/collections/asr/models/ctc_bpe_models.py [hunk 1]
```diff
@@ -114,22 +118,67 @@ def get_vocab():
         self._wer = WERBPE(tokenizer=self.tokenizer, batch_dim_index=0, use_cer=False, ctc_decode=True)
 
     def transcribe(self, paths2audio_files: List[str], batch_size: int = 4) -> List[str]:
```
Comment (okuchaiev) [inline nemo/collections/asr/models/ctc_bpe_models.py:120]: This looks awfully similar to EncDecCTCModel.transcribe(...)  My understanding that the only difference is the type f self._wer (which is WERBPE in this case).   Can we move this method to ASRModel then?

File: nemo/core/classes/modelPT.py [hunk 2]
```diff
@@ -186,14 +186,24 @@ def restore_from(cls, restore_path: str):
...
-            cls.__unpack_nemo_file(path2file=restore_path, out_folder=tmpdir)
-            config_yaml = path.join(tmpdir, _MODEL_CONFIG_YAML)
-            model_weights = path.join(tmpdir, _MODEL_WEIGHTS)
-            conf = OmegaConf.load(config_yaml)
-            OmegaConf.set_struct(conf, True)
-            instance = cls.from_config_dict(config=conf)
-            instance.load_state_dict(torch.load(model_weights))
+            try:
+                cls.__unpack_nemo_file(path2file=restore_path, out_folder=tmpdir)
+
```
Comment (okuchaiev) [inline nemo/core/classes/modelPT.py:194]: why wrap it with empty lines up and down?

File: nemo/collections/asr/models/ctc_bpe_models.py (hunk 1 above)
Comment (titu1994) [reply nemo/collections/asr/models/ctc_bpe_models.py:120]: Entire the same, but its signature to the dataloader setup is different (doesnt have labels, uses tokenizer instead). Any idea about how that should be refactored? Maybe a method which does the general steps but takes a dictionary topass to the dataloader?

File: nemo/core/classes/modelPT.py (hunk 2 above)
Comment (titu1994) [reply nemo/core/classes/modelPT.py:194]: Oh for clarity, will revert.

File: nemo/collections/asr/models/ctc_bpe_models.py (hunk 1 above)
Comment (titu1994) [reply nemo/collections/asr/models/ctc_bpe_models.py:120]: Ive tried adding an internal (semi private) method - `def _setup_transcribe_dataloader(self, config: Dict) -> DataLoader:` which can be overriden by subclass.

File: nemo/core/classes/modelPT.py (hunk 2 above)
Comment (titu1994) [reply nemo/core/classes/modelPT.py]: Done

File: Jenkinsfile [hunk 3]
```diff
@@ -162,6 +162,28 @@ pipeline {
       }
     }
 
+    stage('L2: Speech to Text WPE') {
```
Comment (blisc) [inline Jenkinsfile:165]: Please parallel this with another test of similar duration

Comment (titu1994) [reply Jenkinsfile:165]: BPE models take more memory, but I might be able to put it on GPU 1 alongside speaker embeddings and speech recognition. I'll try.

Comment (titu1994) [reply Jenkinsfile]: @blisc Done
